File: src/utils/husky-library.ts
// Types
import type { PackageJsonScript } from '@/types/package-json'
import type { PackageManager } from '@/types/package-manger'

// Utils
import { writeMessage } from './console'
import { installDependencies } from './dependencies'
import { getErrorMessage } from './errors'
import { modifyNpmIgnore } from './npm'
import { addScript } from './package-json'
import { createFolder, exists } from './user-os'

// NodeJS
import fs from 'node:fs/promises'
import process from 'node:process'

// Constants
import { UTF8_ENCODING } from '@/constants/encoding'
import { LINT_STAGED_CONFIG } from '@/constants/lint-staged'

interface Props {
  packageManagerToUse: PackageManager
  packageJsonPath: string
  useCommitlint: boolean
  shouldPublishToNpm: boolean
}

type AddNecessaryScriptsToPackageJsonProps = Omit<Props, 'useCommitlint'>
type CreateHuskyConfigFiles = Pick<
  Props,
  'packageManagerToUse' | 'useCommitlint'
>

export async function generateHuskyConfig({
  packageManagerToUse,
  packageJsonPath,
  useCommitlint,
  shouldPublishToNpm
}: Props) {
  try {
    writeMessage({
      type: 'config',
      message: "Generating Husky's configuration..."
    })

    await Promise.all(
      [
        installDependencies({
          packageManagerToUse,
          packagesToInstall: 'husky'
        }),

        createHuskyConfigFiles({ packageManagerToUse, useCommitlint }),

        addNecessaryScriptsToPackageJson({
          packageJsonPath,
          packageManagerToUse,
          shouldPublishToNpm
        }),

        shouldPublishToNpm ? modifyNpmIgnore('.husky') : null
      ].filter(promise => promise != null)
    )

    writeMessage({
      type: 'success',
      message: "Husky's configuration generated successfully"
    })
  } catch {
    writeMessage({
      type: 'error',
      message: getErrorMessage('Husky')
    })

    process.exit(1)
  }
}

async function createHuskyConfigFiles({
  packageManagerToUse,
  useCommitlint
}: CreateHuskyConfigFiles) {
  writeMessage({
    type: 'info',
    message: 'Creating configuration file...'
  })

  const existsHuskyFolder = await exists(`${process.cwd()}/.husky`)

  if (!existsHuskyFolder) {
    await createFolder('.husky')
  }

  const preCommitFileValue = useCommitlint
    ? LINT_STAGED_CONFIG[packageManagerToUse]
    : `${packageManagerToUse} test`

  await fs.writeFile('.husky/pre-commit', preCommitFileValue, {
    encoding: UTF8_ENCODING
  })

  writeMessage({
    type: 'success',
    message: 'Configuration file (pre-commit) created successfully'
  })
}

async function addNecessaryScriptsToPackageJson({
  packageJsonPath,
  packageManagerToUse,
  shouldPublishToNpm
}: AddNecessaryScriptsToPackageJsonProps) {
  const huskyScriptsForYarn: PackageJsonScript | PackageJsonScript[] =
    !shouldPublishToNpm
      ? { key: 'postinstall', value: 'husky' }
      : [
          { key: 'postinstall', value: 'husky' },
          { key: 'prepack', value: 'pinst --disable' },
          { key: 'postpack', value: 'pinst --enable' }
        ]

  const scriptsToAdd: PackageJsonScript | PackageJsonScript[] =
    packageManagerToUse !== 'yarn'
      ? { key: 'prepare', value: 'husky' }
      : huskyScriptsForYarn

  await addScript({
    packageJsonPath,
    scriptsToAdd
  })
}

File: internal/utils/husky-library.go
package utils

import (
	"github.com/RaulCatalinas/HuskyBC/internal/constants"
	"github.com/RaulCatalinas/HuskyBC/internal/enums"
)

type Props struct {
	PackageManagerToUse enums.PackageManager
	PackageJsonPath     string
	UseCommitlint       bool
	ShouldPublishToNpm  bool
}

func GenerateHuskyConfig(props Props) {
	WriteMessage(WriteMessageProps{
		Type:    enums.MessageTypeConfig,
		Message: "Generating Husky's Configuration...",
	})

	InstallDependencies(InstallProps{
		PackageManagerToUse: props.PackageManagerToUse,
		PackagesToInstall:   []string{"husky"},
	})

	createHuskyConfigFiles(props.PackageManagerToUse, props.UseCommitlint)

	addNecessaryScriptsToPackageJson(
		props.PackageJsonPath,
		props.PackageManagerToUse,
		props.ShouldPublishToNpm,
	)

	if props.ShouldPublishToNpm {
		modifyNpmIgnore(".husky")
	}

	WriteMessage(WriteMessageProps{
		Type:    enums.MessageTypeSuccess,
		Message: "Husky's configuration generated successfully",
	})
}

func createHuskyConfigFiles(packageManagerToUse enums.PackageManager, useCommitlint bool) {
	WriteMessage(WriteMessageProps{
		Type:    enums.MessageTypeInfo,
		Message: "Creating configuration file...",
	})

	CheckinFolderOrFile(constants.PATH_DIR_HUSKY, true)

	var preCommitFileValue string

	if useCommitlint {
		preCommitFileValue = constants.LINT_STAGED_CONFIG[enums.PackageManager(packageManagerToUse)]
	} else {
		preCommitFileValue = string(packageManagerToUse) + "test"
	}

	writeFile(constants.PATH_DIR_HUSKY+"/pre-commit", []byte(preCommitFileValue))

	WriteMessage(WriteMessageProps{
		Type:    enums.MessageTypeSuccess,
		Message: "Configuration file (pre-commit) created successfully",
	})
}

type packageJsonScript struct {
	Key   string
	Value string
}

func addNecessaryScriptsToPackageJson(
	packageJsonPath string,
	packageManagerToUse enums.PackageManager,
	shouldPublishToNpm bool,
) {
	WriteMessage(WriteMessageProps{
		Type:    enums.MessageTypeInfo,
		Message: "Adding necessary scripts to package.json...",
	})

	var huskyScriptsForYarn interface{}
	var scriptsToAdd interface{}

	if shouldPublishToNpm {
		huskyScriptsForYarn = packageJsonScript{Key: "postintall", Value: "husky"}
	} else {
		huskyScriptsForYarn = []packageJsonScript{
			{Key: "postinstall", Value: "husky"},
			{Key: "prepack", Value: "pinst --disable"},
			{Key: "postpack", Value: "pinst --enable"},
		}
	}

	if packageManagerToUse != "yarn" {
		scriptsToAdd = packageJsonScript{Key: "prepare", Value: "husky"}
	} else {
		scriptsToAdd = huskyScriptsForYarn
	}

	addScript(addScriptProps{
		ScriptsToAdd:    scriptsToAdd,
		PackageJsonPath: packageJsonPath,
	})
}
